[ADD] [16.0] sale_stock_expiry_date_on_qty_at_date_widget

Shide opened the pull request:
Add a predicted expiration date on Qty at Date Widget.
Also shows a button to see Quants and a Kanban view to see Quants.

MT-8827 @moduon @rafaelbn @yajo @fcvalgar @Gelojr please review if you want 😄

Shide committed:
[ADD] sale_stock_expiry_date_on_qty_at_date_widget

rafaelbn commented on sale_stock_expiry_date_on_qty_at_date_widget/README.rst:
```diff
@@ -0,0 +1,122 @@
+Usage
+=====
+
+To use this module, you need to:
```
I have to install sale_management because in runboat it wasn't installed. So I guess there is not dependency 😄 just a comment. Maybe nothing to FIX

rafaelbn reviewed (approved):
QA 👍🏼 

https://www.loom.com/share/ad97763ac2734aea96f8e814ccfca7e8?sid=cd27ee82-ba09-4d27-998a-81b9fc969f1a

yajo commented on sale_stock_expiry_date_on_qty_at_date_widget/readme/CONTEXT.md:
```diff
@@ -0,0 +1,4 @@
+This module was developed because sometimes Commercial users want to know where is the Expiraiton Date of the products.
```
```suggestion
This module was developed because sometimes salesperson users need to know when products will expire.
```

yajo commented on sale_stock_expiry_date_on_qty_at_date_widget/readme/CONTEXT.md:
```diff
@@ -0,0 +1,4 @@
+This module was developed because sometimes Commercial users want to know where is the Expiraiton Date of the products.
+
+It will be useful for you if your Commercial users need to share this informations with its customers.
```
```suggestion
It will be useful for you if your salespersons need to share this information with their customers.
```

yajo commented on sale_stock_expiry_date_on_qty_at_date_widget/readme/USAGE.md:
```diff
@@ -0,0 +1,13 @@
+To use this module, you need to:
+
+1. Install sale_management module
```
issue: this step doesn't belong to this section. Usage section is for end users. Instead, put it in the `INSTALL.md` file.

yajo commented on sale_stock_expiry_date_on_qty_at_date_widget/readme/USAGE.md:
```diff
@@ -0,0 +1,13 @@
+To use this module, you need to:
+
+1. Install sale_management module
+2. Go to Inventory > Settings > Traceability and enable Expiration Dates
```
issue: this step should belong to the `CONFIGURE.md` section.

suggestion: if you always prefix these with `1.` instead of `2., 3., ...`, then it's easier to insert new steps before or after without changing the rest of steps.

yajo commented on sale_stock_expiry_date_on_qty_at_date_widget/readme/USAGE.md:
```diff
@@ -0,0 +1,13 @@
+4. Fill stock of this product with Lots and Expiration Dates
+5. Enable an user with only this permissions:
+  - Sales: User: Own Documents Only
+  - Technical: Manage Lots / Serial Numbers
```
issue: also to configure section.

yajo commented on sale_stock_expiry_date_on_qty_at_date_widget/readme/USAGE.md:
```diff
@@ -0,0 +1,13 @@
+6. Create a new Sale and select the product you've configured previously
+7. Check the Widget that shows info about the stock
+8. Check the next Expiration Date is correctly set
+9. Click on View Quants to see all Quants and its Expiration Dates
```
suggestion: I think a screenshot would help here.

yajo commented on sale_stock_expiry_date_on_qty_at_date_widget/static/src/widgets/qty_at_date_widget.js:
issue: suffix this file with `.esm.js`, so ESLint knows it's a module, and lints it appropriately. Relevant config: https://github.com/OCA/sale-workflow/blob/1593a90370898a879957d49bd9a991148539b022/.eslintrc.yml#L11-L16

yajo commented on sale_stock_expiry_date_on_qty_at_date_widget/static/src/widgets/qty_at_date_widget.xml:
```diff
@@ -0,0 +1,39 @@
+<!-- Copyright 2025 Moduon Team S.L. <info@moduon.team>
+     License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl). -->
+<template id="template" xml:space="preserve">
+    <!-- TODO: rename this to QtyAtDatePopup in master version -->
```
suggestion: I think this comment will be more helpful for your future self like this:

```suggestion
    <!-- TODO: rename this to QtyAtDatePopup in Odoo 19 -->
```

yajo commented on sale_stock_expiry_date_on_qty_at_date_widget/tests/test_predicted_expiry_date.py:
```diff
@@ -0,0 +1,101 @@
+        )
+        now = datetime.now()
+        cls.high_exp_date = now + timedelta(days=10)
+        cls.lower_exp_date = now + timedelta(days=5)
```
issue: since you're using time-aware tests, it's a good practice to use freezegun and and hardcode expected dates. Makes it easier to reason about it. 

It also saves you from random problems when running tests in different timezones where the start and end datetime of the test can change the date, just for milliseconds.

More details: https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst#avoid-dynamic-dates

yajo commented on sale_stock_expiry_date_on_qty_at_date_widget/views/sale_order_line_view.xml:
```diff
@@ -0,0 +1,38 @@
+<!-- Copyright 2025 Moduon Team S.L. <info@moduon.team>
+     License LGPL-3.0 or later (https://www.gnu.org/licenses/lgpl). -->
+<odoo>
+    <record id="view_order_form_inherit_sale_stock_qty" model="ir.ui.view">
+            <field
+            name="name"
+        >sale.order.line.tree.sale.stock.predicted.expiration.date</field>
+            <field
+            name="inherit_id"
+            ref="sale_stock.view_order_form_inherit_sale_stock_qty"
+        />
+            <field name="model">sale.order</field>
+            <field name="arch" type="xml">
+                <xpath
+                expr="//page/field[@name='order_line']/form//field[@name='company_id']"
+                position="after"
+            >
+                    <!-- below fields are used in the widget qty_at_date_widget -->
+                    <field name="predicted_first_expiration_date" invisible="1" />
+                </xpath>
+                <xpath
+                expr="//page/field[@name='order_line']/tree//field[@name='company_id']"
+                position="after"
+            >
+                    <!-- below fields are used in the widget qty_at_date_widget -->
+                    <field name="predicted_first_expiration_date" invisible="1" />
+                </xpath>
+                <xpath
+                expr="//page/field[@name='order_line']/kanban//field[@name='company_id']"
+                position="after"
+            >
+                    <!-- below fields are used in the widget qty_at_date_widget -->
+                    <field name="predicted_first_expiration_date" invisible="1" />
+                </xpath>
+            </field>
+        </record>
```
suggestion: indentation looks a bit weird in this file. Prettier is whitespace-conservative, so its reformats are sometimes not what you'd expect. Would you mind to prettify it a bit please?

yajo reviewed (changes requested):
It looks amazing, thanks!

I added a few comments, but they are mostly about cosmetic stuff and docs. However, some are issues.

yajo commented:
Suggestion for a slightly shorter technical module name: `sale_stock_qty_at_expiry_date`

fcvalgar reviewed (approved):
Functional review, 

LGTM, thank you man @Shide 

![image](https://github.com/user-attachments/assets/69ce8a82-f759-4091-994a-aeb28c1f347a)


![image](https://github.com/user-attachments/assets/24facc67-52c2-439a-8f30-8a335539fde8)

Gelojr reviewed (approved):
LGTM. Nice @Shide. Is the  predicted expiration date of the oldest quant, isn't it?

yajo reviewed (approved):
/ocabot merge nobump

yajo commented:
/ocabot merge nobump

Shide commented:
> LGTM. Nice @Shide. Is the predicted expiration date of the oldest quant, isn't it?

Is the expiration date of the quant (lot) that will be used if you confirm the order and reserve the quantities in that moment.
The reservation method set on the Category of the Product is taken in consideration (FIFO, FEFO, etc)

rafaelbn commented:
/ocabot merge patch

rafaelbn commented:
/ocabot rebase

Shide committed:
[FIX] *_triple_discount, *_fixed_discount rebel module groups

Added 2 groups of rebel_module_groups:
- sale_order_general_discount_triple,sale_pricelist_triple_discount,sale_triple_discount
- sale_fixed_discount

rafaelbn commented:
/ocabot merge patch
